java/arcs/android/storage/service/StorageServiceNgImpl.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.storage.service
+
+import android.os.Parcel
+import androidx.annotation.VisibleForTesting
+import arcs.android.storage.readStoreOptions
+import arcs.core.storage.ActiveStore
+import arcs.core.storage.DriverFactory
+import arcs.core.storage.StorageKey
+import arcs.core.storage.UntypedActiveStore
+import arcs.core.storage.UntypedProxyMessage
+import arcs.core.storage.WriteBackProvider
+import java.util.concurrent.ConcurrentHashMap
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+/**
+ * Implementation of the [IStorageServiceNg] AIDL interface. Responsible for forwarding messages
+ * to [ActiveStore]s and back again.
+ */
+class StorageServiceNgImpl(
+  private val scope: CoroutineScope,
+  private val driverFactory: DriverFactory,
+  private val writeBackProvider: WriteBackProvider,
+  private val devToolsProxy: DevToolsProxyImpl?,
+  /** Callback to trigger when a proxy message has been received and sent to the store. */
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit
+) : IStorageServiceNg.Stub() {
+  // TODO(b/162747024): Replace this with an LruCache so its size doesn't grow unbounded.
+  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
```

Comment:
I would argue that we consider ways to remove this completely. The internal stores map is an implementation detail.

If we're not ready to give up the assertions on the contents of the map, perhaps a better approach for now might be to leave this property private, and expose a "active storageKeys" property that can remain public; this may actually be useful for other purposes later, anyway (for logging or management objects, for example). 

i.e. (assuming we also moved to using a mutex)

```
fun activeStorageKeys(): Collection<StorageKey> = storesMutex.withLock { stores.keys }
```

Reply:
(This also pertains to several other comments below about not testing implementation details)

I'm not sure that the stores cache is strictly an internal implementation detail. Or rather, I think it is, but it's also a really important detail that warrants testing _somewhere_. Perhaps not necessarily here, but then where else?

The reason the cache is so important is not just to save memory (though that is important too), but also for latency reasons. We want all store ops in the storage service to be routed through the same Store object, so that updates can be propagated out quickly to their listeners. I'm fairly sure that our architecture guarantees that if you were to create separate stores for the same storage key, the others stores would hear the updates eventually, and then invoke their callbacks. But I'm not certain how long that would take to propagate, and possibly the delay depends on the specific storage driver used, and really we just want to avoid the whole thing entirely.

With that in mind, testing the caching behaviour here actually seems like the simplest and most accurate thing to me. Some alternatives I can think of are:

1. Instead of checking the `stores` cache, check the `.store` field of each storage channel. This is just kicking the can down the road, and testing another private field.

2. Try to test the expected behaviour instead. e.g. instead of checking that the store was correctly cached, check that the time delay between writing via one channel and being updated via another channel is very small. That test sounds very like it will be very flaky, and I worry that different storage drivers might exhibit different characteristics (e.g. RamDisk might always be fast enough, but SQLite mightn't). This also doesn't test that we actually _cached_ the instance, so the memory usage concerns would still be untested.

3. Instead of using a `ConcurrentHashMap` directly for the store cache, build a `StoreCache` abstraction around it, and then inject a factory for it so that we can mock it out in the test. Now, rather than inspecting the store cache itself, we can verify that particular methods were called on the mock (e.g. `.create()` called the first time, `.get()` called the second time). 

I'm curious if either of you have other ideas.

The main question I have is: when you make particular implementation choices for (hopefully well-founded) performance reasons (whether memory or latency), what is the best way to ensure that they continue to work over time, if not by unit testing them?

Reply:
For now, I'd be happy if we just keep the stores property private, and create a derived property that exposes only the storageKeys. As far as I can tell, this satisfies the requirements of the testing. And that property seems genuinely useful and safe, and could be exposed without the @VisibleForTesting annotation.

Longer term, I think your suggestion 3 feels like the best solution, I think that's where I'd have eventually ended up.

I think we should just all pretend #2 was never even an option ;-)

Reply:
#1 Actually doesn't sound like a bad idea. And I think it's comes closer to confirming our desired result: that the two endpoints are referencing the same store.

Reply:
I agree with @csilvestrini, I think it's worthwhile checking stores are created as expected. I've added the public method `activeStorageKeys()` which returns the collection of `StorageKey`. (Note that I haven't guarded this with a mutex yet. There's a lot going on in this PR so I've filed a bug and I plan to fix concurrency problems in a follow up PR). 

As per @jblebrun comment [here](https://github.com/PolymerLabs/arcs/pull/6603#discussion_r527284135) , I think it's probably also important to check that the channels are communicating with the intended stores. So I've also made `storageKey` a new public attribute in the `StorageChannelImpl` class. Let me know what you think.

Reply:
If we really want to check that it's the same store, is checking storage keys sufficient? It's possible to create two store instances for the same key, so shouldn't the test here to actually ensure that the store instance itself is the same?